File: bindings/dotnet/src/Wcl/Eval/BlockRef.cs
using System.Collections.Generic;
using System.Linq;
using Wcl.Core;

namespace Wcl.Eval
{
    public class BlockRef
    {
        public string Kind { get; set; }
        public string? Id { get; set; }
        public OrderedMap<string, WclValue> Attributes { get; set; }
        public List<BlockRef> Children { get; set; }
        public List<DecoratorValue> Decorators { get; set; }

        public BlockRef(string kind, string? id,
                        OrderedMap<string, WclValue> attributes, List<BlockRef> children,
                        List<DecoratorValue> decorators)
        {
            Kind = kind; Id = id;
            Attributes = attributes; Children = children;
            Decorators = decorators;
        }

        public bool HasDecorator(string name) => Decorators.Any(d => d.Name == name);

        public DecoratorValue? GetDecorator(string name) =>
            Decorators.FirstOrDefault(d => d.Name == name);

        public WclValue? Get(string key) =>
            Attributes.TryGetValue(key, out var val) ? val : null;
    }

    public class DecoratorValue
    {
        public string Name { get; set; }
        public OrderedMap<string, WclValue> Args { get; set; }

        public DecoratorValue(string name, OrderedMap<string, WclValue> args)
        {
            Name = name;
            Args = args;
        }
    }
}

File: bindings/jvm/src/main/java/io/github/wiltaylor/wcl/eval/BlockRef.java
package io.github.wiltaylor.wcl.eval;

import java.util.LinkedHashMap;
import java.util.List;

public final class BlockRef {
    private final String kind;
    private final String id;
    private final LinkedHashMap<String, WclValue> attributes;
    private final List<BlockRef> children;
    private final List<DecoratorValue> decorators;

    public BlockRef(String kind, String id,
                    LinkedHashMap<String, WclValue> attributes,
                    List<BlockRef> children, List<DecoratorValue> decorators) {
        this.kind = kind;
        this.id = id;
        this.attributes = attributes;
        this.children = children;
        this.decorators = decorators;
    }

    public String getKind() { return kind; }
    public String getId() { return id; }
    public LinkedHashMap<String, WclValue> getAttributes() { return attributes; }
    public List<BlockRef> getChildren() { return children; }
    public List<DecoratorValue> getDecorators() { return decorators; }

    public WclValue get(String key) {
        return attributes.get(key);
    }

    public boolean hasDecorator(String name) {
        return decorators.stream().anyMatch(d -> d.name().equals(name));
    }

    public DecoratorValue getDecorator(String name) {
        return decorators.stream()
                .filter(d -> d.name().equals(name))
                .findFirst()
                .orElse(null);
    }
}
